File: mapwidget/cesium.py
import os
import pathlib
import anywidget
import traitlets


class Map(anywidget.AnyWidget):
    """Map widget

    Args:
        anywidget (_type_): _description_
    """

    _cwd = os.path.dirname(os.path.abspath(__file__))
    _esm = pathlib.Path(os.path.join(_cwd, 'javascript', 'cesium.js'))
    _css = pathlib.Path(os.path.join(_cwd, 'styles', 'cesium.css'))
    default_token = os.environ.get('CESIUM_TOKEN')
    token = traitlets.Unicode(default_token).tag(sync=True)
    center = traitlets.List([40, -100]).tag(sync=True, o=True)
    zoom = traitlets.Float(4).tag(sync=True, o=True)
    bounds = traitlets.List([0, 0, 0, 0]).tag(sync=True, o=True)
    width = traitlets.Unicode('100%').tag(sync=True, o=True)
    height = traitlets.Unicode('600px').tag(sync=True, o=True)
    clicked_latlng = traitlets.List([None, None]).tag(sync=True, o=True)
    altitude = traitlets.Float(0).tag(sync=True, o=True)

    # czml = traitlets.Tuple().tag(sync=True, o=True)
    # kml_url = traitlets.Unicode().tag(sync=True, o=True)
    # geojson = traitlets.Unicode().tag(sync=True, o=True)
    # animation = traitlets.Bool(True).tag(sync=True, o=True)
    # base_layer_picker = traitlets.Bool(True).tag(sync=True, o=True)
    # geocoder = traitlets.Bool(True).tag(sync=True, o=True)
    # home_button = traitlets.Bool(True).tag(sync=True, o=True)
    # infobox = traitlets.Bool(True).tag(sync=True, o=True)
    # scene_mode_picker = traitlets.Bool(True).tag(sync=True, o=True)
    # selection_indicator = traitlets.Bool(True).tag(sync=True, o=True)
        
    timeline = traitlets.Bool(True).tag(sync=True, o=True)
    # navigation_help_button = traitlets.Bool(True).tag(sync=True, o=True)
    # navigation_instructions_initially_visible = traitlets.Bool(False).tag(sync=True, o=True)
    # scene_3D_only = traitlets.Bool(False).tag(sync=True, o=True)
    # scene_mode = traitlets.CaselessStrEnum(['COLUMBUS_VIEW', 'SCENE2D', 'SCENE3D'],
    #                              default_value='SCENE3D',
    #                              allow_none=False).tag(sync=True, o=True)
    # enable_lighting = traitlets.Bool(False).tag(sync=True, o=True)

File: mapwidget/javascript/cesium.js
function loadScript(src) {
    return new Promise((resolve, reject) => {
        let script = Object.assign(document.createElement("script"), {
            type: "text/javascript",
            async: true,
            src: src,
        });
        script.addEventListener("load", resolve);
        script.addEventListener("error", reject);
        document.body.appendChild(script);
    });
}

await loadScript(
    "https://cesium.com/downloads/cesiumjs/releases/1.103/Build/Cesium/Cesium.js"
);

export function render(view) {
    // Header
    let center = view.model.get("center");
    let altitude = view.model.get("altitude");

    let width = view.model.get("width");
    let height = view.model.get("height");

    let timeline = view.model.get('timeline');
    // let animation = view.model.get('animation');
    // let baseLayerPicker = view.model.get('base_layer_picker');
    // let geocoder = view.model.get('geocoder');
    // let homeButton = view.model.get('home_button');
    // let infoBox = view.model.get('infobox');
    // let sceneModePicker = view.model.get('scene_mode_picker');
    // let selectionIndicator = view.model.get('selection_indicator');
    // let navigationHelpButton = view.model.get('navigation_help_button');
    // let navigationInstructionsInitiallyVisible = view.model.get('navigation_instructions_initially_visible');
    // let scene3DOnly = view.model.get('scene_3D_only');
    // let sceneMode_name = view.model.get('scene_mode');


    const div = document.createElement("div");
    div.style.width = width;
    div.style.height = height;

    Cesium.Ion.defaultAccessToken = view.model.get("token");

    // Map content
    const viewer = new Cesium.Viewer(div, {
        terrainProvider: Cesium.createWorldTerrain(),
    });

    //const buildingTileset = viewer.scene.primitives.add(
    //    Cesium.createOsmBuildings()
    //);

    viewer.camera.flyTo({
        destination: Cesium.Cartesian3.fromDegrees(
            center[1],
            center[0],
            altitude
        ),
        orientation: {
            heading: Cesium.Math.toRadians(0.0),
            pitch: Cesium.Math.toRadians(-15.0),
        },
    });

    viewer.camera.moveEnd.addEventListener(function () {
        var cameraPosition = viewer.camera.position;
        var center =
            Cesium.Ellipsoid.WGS84.cartesianToCartographic(cameraPosition);
        var lon = Cesium.Math.toDegrees(center.longitude);
        var lat = Cesium.Math.toDegrees(center.latitude);
        var height = viewer.camera.positionCartographic.height;
        var zoomLevel = 18 - Math.log2(height);
        view.model.set("center", [lat, lon]);
        view.model.set("zoom", zoomLevel);
        view.model.save_changes();
    });

    // Footer
    view.el.appendChild(div);
}
